File: python/luhn/luhn.py
class Luhn(object):
    def __init__(self, card_num):
        # normalizing input
        try:
            self.cn = [int(c) for c in card_num.replace(" ", "")][::-1]
        except ValueError:
            self.cn = [0]

    def is_valid(self):
        if len(self.cn) < 2:
            return False

        # sum of the doubles
        odds = sum(
            sum(divmod(i * 2, 10))
            for i in self.cn[1::2]
        )
        evens = sum(self.cn[0::2])

        return (odds + evens) % 10 == 0

File: php/luhn/luhn.php
<?php

function isValid(string $card_num): bool
{
    # normalizing input
    $nospace = preg_replace("/\s/", "", $card_num);
    $digits = strrev(preg_replace("/\D/", "", $card_num));

    if (
        (strlen($digits) < 2) # must be at least 2 digits
         || (strlen($digits) != strlen($nospace)) # must be digits only
    ) {
        return false;
    }

    # init the check sum
    $checksum = 0;
    foreach (str_split($digits) as $k => $i) {
        switch ($k % 2) {
            case 0: # even positions
                $add = $i;
                break;
            case 1: # odd positions
                $double = $i * 2;
                $add = ($double > 9) ? ($double - 9) : $double; # if double > 9
                break;
        }
        $checksum += $add;
    }

    return ($checksum % 10) == 0;
}

if (!debug_backtrace()) {
    isValid("091");
}
